proxy-server/proxy/config.go

```diff
@@ -0,0 +1,202 @@
+package proxy
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	kubecache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type ProxyConfig struct {
+	// Map of requested domain prefix to k8s dst address
+	TargetMap map[string]string
+	mutex     sync.Mutex
+}
+
+/*
+	Get service address for proxy target
+*/
+func (pc *ProxyConfig) GetTarget(key string) string {
+	pc.mutex.Lock()
+	defer pc.mutex.Unlock()
+	target, ok := pc.TargetMap[key]
+	if !ok {
+		return ""
+	}
+	return target
+}
+
+/*
+	Load proxy target map from a ConfigMap
+*/
+func (pc *ProxyConfig) Load(configmap interface{}) error {
+	pc.mutex.Lock()
+	defer pc.mutex.Unlock()
+
+	data, err := loadConfigMap(configmap)
+	if err != nil {
+		return err
+	}
+
+	// Check for changes to avoid spamming logs with "Loaded proxy config"
+	targetsUpdated := len(pc.TargetMap) != len(data)
```

Comment:
This makes me nervous, but I _think_ it's ok - are we guaranteed to get every event in order from the config watcher?

Reply:
Here's a good rundown of the SharedInformer pattern that is used here: https://engineering.bitnami.com/articles/a-deep-dive-into-kubernetes-controllers.html

Summary is, events are placed on a workqueue. If multiple events for the same resource are on the queue they get collapsed down to the latest version. There's no 100% guarantee that there wasn't a problem with an event (error processing, missed by k8s controller, lost in network, etc.), however that's where the "Resync Period" comes into play. I have it set to 30 seconds, so regardless of what happened with events in that time it will _always_ call the `UpdateFunc` with the latest version of the object ever 30s. And that's why I did the work to check if there was actually a change in the configmap since the last update, to avoid filling the logs up with "blah blah config updated" every 30s when there wasn't necessarily a change.